Document dra-driver-nvidia-gpu with KubeVirt VFIO GPU passthrough (#1218)
kubernetes-sigs/dra-driver-nvidia-gpu: site/content/docs/guides/kubevirt-vfio-gpu-passthrough.md

---
title: KubeVirt VFIO GPU passthrough
linkTitle: KubeVirt VFIO GPU passthrough
weight: 40
description: Configure the DRA driver for NVIDIA GPUs for KubeVirt VFIO passthrough.
---

> KubeVirt can attach VFIO GPU devices to virtual machines through Kubernetes Dynamic Resource Allocation (DRA). This guide covers the NVIDIA DRA driver configuration needed for KubeVirt passthrough workloads.

Reviewer: "KubeVirt can attach GPU (either in VFIO or mdev) to VirtualMachines through..."

Reviewer 2: @alaypatel07 We don't have vgpu support yet. So I'm not sure we want to call out mdev here. We can add them in the future

Author: Yes, I agree. Left it as is for now.

For KubeVirt feature gates, VMI fields, and VM examples, see the [KubeVirt user guide](https://kubevirt.io/user-guide/).

## Prerequisites

> - Meet the general [driver prerequisites](../prerequisites.md).

Reviewer 2: ```suggestion
- Meet the general driver [prerequisites](../prerequisites.md).
```

> - NVIDIA DRA driver v0.4.0 or later, which supports KEP-5304 device metadata.

Reviewer 3: nit/suggestion add a link to the KEP-5304

> - The driver is installed with `PassthroughSupport` and `DeviceMetadata` enabled.

Reviewer 3: ```suggestion
- The driver is installed with `PassthroughSupport` and `DeviceMetadata` feature gates enabled.
```

> - Use **NVIDIA DRA driver v0.4.0 or later**, which supports KEP-5304 device metadata.

Reviewer 2: Can you link it to the kep here?

Author: Updated.

> - Install the driver with **`PassthroughSupport`** and **`DeviceMetadata`** featuregates enabled.

Reviewer 2: Can you link it to the constraints here so they know what FGs are supported with these 2? https://dra-driver-nvidia-gpu.sigs.k8s.io/docs/reference/feature-gates/#constraints

- `vfio-pci` is available on GPU nodes.
- VFIO device nodes are available under `/dev/vfio`.

## Install the NVIDIA DRA driver

Install the driver with passthrough support and device metadata:

```bash
helm upgrade -i dra-driver-nvidia-gpu oci://registry.k8s.io/dra-driver-nvidia/charts/dra-driver-nvidia-gpu \
    --version 0.4.0 \
    --namespace dra-driver-nvidia-gpu \
    --create-namespace \
    --set resources.gpus.enabled=true \
    --set resources.computeDomains.enabled=false \
    --set gpuResourcesEnabledOverride=true \
    --set nvidiaDriverRoot=/ \
    --set featureGates.PassthroughSupport=true \
    --set featureGates.DeviceMetadata=true
```

Set `nvidiaDriverRoot` based on how the NVIDIA driver is installed on your nodes:

- `/` for a host-installed driver.
- `/run/nvidia/driver` for a GPU Operator-managed driver.
- `/home/kubernetes/bin/nvidia` for a GKE-managed driver.

Verify that the driver registered the expected `DeviceClass` and advertised node resources:

```bash
kubectl get deviceclass

> kubectl get resourceslice -o wide

Reviewer 3: nit: might be good to share what the expected output is here, so its easy to confirm that things installed ok.

```

> ### Stop services and processes using NVIDIA GPUs

Reviewer 3: Is there a specific error users would see that would lead them to thing that one of these services is blocking the dra driver install, like in the logs or seeing the dra driver pods stalled?

Author: Yes, made it clear by adding troubleshooting and references section.

> Before a GPU can be rebound to `vfio-pci`, stop services and workloads that keep open handles on NVIDIA GPU device nodes such as `/dev/nvidia0` and `/dev/nvidia1`.

Reviewer: "Before a GPU can bind to `vfio-pci`..."

Author: isn't rebound fits here, since we are unbinding it from nvidia and binding it to vfio-pci?

Reviewer 2: I agree `rebound` is confusing. It would be better to rephrase it as: "For a GPU to support driver swtiches between nvidia and vfio-pci ..."

Author: Updated.

```bash

> sudo systemctl stop nvidia-dcgm dcgm nvidia-persistenced nvsm

Reviewer: There are a ton of manual steps here, is there a need for a script that does this automatically?

Author: It should be fine for now I think. Whether we need a script really depends on how often people hit these steps and which install modes they’re using, we can add it later if it is repetitve.

```

Verify that no other process holds an open handle on GPU devices:

```bash
sudo lsof /dev/nvidia[^-]
```

This command should produce no output.

> Common processes that block GPU passthrough include:

Reviewer 4: We do support running in a mixed mode, where some GPUs are allocated to sandbox workloads via passthrough while others are used directly by workload containers.

In the upcoming v0.5.0 release, we're adding support to enforce passthrough-only mode on a node. However, that change is not available yet. Until then, mixed workloads can still be scheduled, and in those scenarios, these services must remain running.

@varunrsekar, could you also help document the limitations for this mixed-mode configuration?

Reviewer 2: @shivamerla Is there a problem in the wordings here?
```
- **nvidia-persistenced** — Disabling this service is optional. The DRA driver can handle it automatically.
- **nvidia-dcgm/dcgm** — Disabling this service is optional when running driver v4.5.0 or later.
```

Both of these services mention that its optional to disable. The rest of the services have to be turned off.

Reviewer 2: @Sreeja1725 I feel we shouldn't have to provide instructions on disabling them. We can simply call out the common services and ask the users to make sure they aren't running or are compatible. Reason being that there are 3 different modes of installation:
1. direct on host
2. through gpu operator
3. manual deployment

- **Xorg** — Identify and disable the display manager:

```bash
  sudo systemctl status display-manager
  sudo systemctl disable <display-manager-name>
```

> - **vectorAdd** — A sample CUDA application; stop the process if it is running.

Reviewer 2: we dont need to call this out.

- **nvidia-device-plugin** — Disable or uninstall the legacy NVIDIA device plugin.
- **dcgm-exporter** — Disable or uninstall DCGM exporter.
- **NVIDIA GPU Operator** — If deployed via Helm, its pods may run `nvidia-device-plugin` and `dcgm-exporter`. Disable or uninstall the operator on GPU nodes where you use DRA passthrough.
- **nvidia-persistenced** — Disabling this service is optional. The DRA driver can handle it automatically.

> - **nvidia-dcgm/dcgm** — Disabling this service is optional when running driver v4.5.0 or later.

Reviewer 3: ```suggestion
- **nvidia-dcgm/dcgm** — Disabling this service is optional when running DCGM v4.5.0 or later.
```
for clarity

---
title: KubeVirt VFIO GPU passthrough
linkTitle: KubeVirt VFIO GPU passthrough
weight: 50
description: Configure the DRA driver for NVIDIA GPUs for KubeVirt VFIO passthrough.
---

KubeVirt can attach VFIO GPU devices to virtual machines through Kubernetes Dynamic Resource Allocation (DRA). This guide covers the NVIDIA DRA driver configuration needed for KubeVirt passthrough workloads.

> For KubeVirt feature gates, VMI fields, and VM examples, see the [KubeVirt user guide](https://kubevirt.io/user-guide/).

Reviewer 2: ```suggestion
For KubeVirt feature gates, VM fields, and VM examples, see the [KubeVirt user guide](https://kubevirt.io/user-guide/).
```
Lets keep our examples for just KubeVirt VMs.

## Feature status

This guide uses two Alpha feature gates, both disabled by default:

| Feature gate | Default | Stage |
|---|---|---|
| `PassthroughSupport` | `false` | Alpha |
| `DeviceMetadata` | `false` | Alpha |

> `DeviceMetadata` requires `PassthroughSupport`, and some gates are mutually exclusive. See the [feature gates reference](../reference/feature-gates.md) and [constraints](../reference/feature-gates.md#constraints) for details.

Reviewer 2: ```suggestion
See the feature gates [reference](../reference/feature-gates.md) and [constraints](../reference/feature-gates.md#constraints) for details.
```

## Prerequisites

- Meet the general driver [prerequisites](../prerequisites.md).
- **IOMMU enabled** on GPU nodes. VFIO passthrough requires IOMMU; the GPU kubelet plugin fails to start with `PassthroughSupport` enabled if IOMMU is off.

> - Use **NVIDIA DRA driver v0.4.0 or later** and install it with the **`PassthroughSupport`** and **`DeviceMetadata`** feature gates enabled.

Reviewer 2: ```suggestion
- Use **NVIDIA DRA driver v0.4.0 or later** with the **`PassthroughSupport`** and **`DeviceMetadata`** feature gates enabled.
```

> - Use **KubeVirt v1.8.0 or later**.

Reviewer 2: Can you mention the feature gate needed?
```
Use **KubeVirt v1.8.0 or later** with the **`GPUsWithDRA`** feature gate enabled.
```

## Limitations and considerations

For a GPU to switch between the `nvidia` and `vfio-pci` drivers, nothing on the host may hold an open handle on that GPU's `/dev/nvidia*` device nodes.

Make sure each of the following is either not using the GPU being prepared, or is configured to release it:

| Component | Requirement |
|---|---|

> | **Display server (Xorg)** | Must not hold the GPU being prepared. |

Reviewer 2: nit: `Must be disabled`

> | **nvidia-device-plugin** | Must not manage the GPU allocated for passthrough. On mixed-mode nodes, it can keep advertising other GPUs. |

Reviewer 4: This is not true, we don't recommend running device-plugin alongside of DRA driver on the same node. The mixed-mode here should be running both container and sandbox workloads on the same machine with both allocations managed by the DRA driver. For simplicity we should add a recommendation in this doc for users to disallow non-sandbox workloads on the node by using taints etc, so they can safely disable any other GPU clients that might be deployed by the GPU operator or externally on these nodes.

Reviewer 2: For `nvidia-device-plugin`, we should say it must be disabled.

Its good to call out the common clients in the table. But we can mention they should be disabled for stability.

In general if we're looking to use passthrough, then the mode of operation does not matter. the above statement always applies.

Author: @shivamerla thanks for pointing it out, I think I misunderstood it. Updated.

> | **NVIDIA GPU Operator** | When deployed via Helm, its pods may run `nvidia-device-plugin` and `dcgm-exporter`; the same requirements as those components apply. On mixed-mode nodes, the pods serving other GPUs can keep running. |

Reviewer 2: we dont need to call out GPU operator.

| **nvidia-persistenced** | On DRA driver v0.4.0 or later, the driver automatically disables persistence mode on the target GPU before rebinding it to `vfio-pci` and restores it afterward, so the service can keep running. Older versions don't have this handling and require stopping the service manually. |

> | **dcgm-exporter** | With DCGM v4.5.0 or later, set `DCGM_EXPORTER_ENABLE_GPU_BIND_UNBIND_WATCH=true` so the exporter reloads and releases a GPU when it is switching between the `nvidia` and `vfio-pci` drivers (poll interval is set by `DCGM_EXPORTER_GPU_BIND_UNBIND_POLL_INTERVAL`, default `1s`). Bind/unbind watching is not supported before DCGM v4.5.0; on those versions, stop the exporter before the GPU is allocated for passthrough. |

Reviewer 2: Lets say it is recommended to disable this service. But add a note saying "DCGM EXPORTER v4.5.0+ has an experimental feature `DCGM_EXPORTER_ENABLE_GPU_BIND_UNBIND_WATCH=true` (plus a poll frequency `DCGM_EXPORTER_GPU_BIND_UNBIND_POLL_INTERVAL=1s`) to automatically release GPUs when it is switching between the nvidia and vfio-pci drivers".  And ask to use this with caution.

> | **dcgm-exporter** | With DCGM v4.5.0 or later, set `DCGM_EXPORTER_ENABLE_GPU_BIND_UNBIND_WATCH=true` so the exporter reloads and releases a GPU when it is switching between the `nvidia` and `vfio-pci` drivers (poll interval is set by `DCGM_EXPORTER_GPU_BIND_UNBIND_POLL_INTERVAL`, default `1s`). Bind/unbind watching is not supported before DCGM v4.5.0; on those versions, stop the exporter before the GPU is allocated for passthrough. |

Reviewer 2: Can you add a row for `dcgm` before `dcgm-exporter`? And similarly mention it is recommended to disable this service. But with a note saying "DCGM v4.5.0+ can automatically release GPUs when it is switching between the nvidia and vfio-pci drivers "

### Creating the claim template

```yaml
apiVersion: resource.k8s.io/v1
kind: ResourceClaimTemplate
metadata:
  name: dra-gpu-claim-template
spec:
  spec:
    devices:
      config:
      - requests:
        - dra-gpu
        opaque:
          driver: gpu.nvidia.com
          parameters:
            apiVersion: resource.nvidia.com/v1beta1
            kind: VfioDeviceConfig
            iommu:
              backendPolicy: LegacyOnly
              enableAPIDevice: true
      requests:
      - name: dra-gpu
        exactly:
          allocationMode: ExactCount
          count: 1
          deviceClassName: vfio.gpu.nvidia.com
```

Save it as `vfio-gpu-rct.yaml`.

### Apply the manifest

```bash
kubectl apply -f vfio-gpu-rct.yaml
```

Example output:

```
resourceclaimtemplate.resource.k8s.io/dra-gpu-claim-template created
```

### VfioDeviceConfig parameters

The opaque `VfioDeviceConfig` block tells the NVIDIA DRA driver which VFIO device nodes to mount into virt-launcher through CDI.

- **`enableAPIDevice: true`** — Mounts the VFIO control device `/dev/vfio/vfio` into the virt-launcher pod. KubeVirt **requires** this device to manage VFIO PCI assignments through libvirt.

> - **`backendPolicy: LegacyOnly`** — Selects the legacy IOMMU VFIO backend (`/dev/vfio/<iommu-group>`). The alternative, `PreferIommuFD`, uses the IOMMUFD backend (`/dev/iommu` and `/dev/vfio/devices/vfio*`) when available on the host.

Reviewer 2: ```suggestion
- **`backendPolicy: LegacyOnly`** — Selects the legacy IOMMU VFIO backend (`/dev/vfio/<iommu-group>`). The alternative, `PreferIommuFD`, uses the IOMMUFD backend (`/dev/vfio/devices/vfio*`) when available on the host.
```

Keep `backendPolicy: LegacyOnly` for KubeVirt, which does not support the IOMMUFD backend yet.

## Troubleshooting

If a VM fails to start, check the kubelet-plugin logs for prepare errors:

```bash
kubectl logs -n dra-driver-nvidia-gpu -l dra-driver-nvidia-gpu-component=kubelet-plugin -c gpus
```

> A prepare error usually means a host process still holds the GPU. On the GPU node, find which process holds a handle on an NVIDIA device:

Reviewer 2: ```suggestion
A `DeadlineExceeded` error for NodePrepareResources for the workload usually means a service still holds the GPU. On the GPU node, find which process holds a handle on an NVIDIA device:
```

```bash
for f in /proc/[0-9]*/fd/*; do t=$(readlink "$f" 2>/dev/null) || continue; case "$t" in /dev/nvidia[0-9]*) echo "PID $(echo "$f" | cut -d/ -f3) holds $t";; esac; done
```

> If the command returns output, note the PID, identify the owning service, and stop those processes. If it returns nothing, no host process is holding the GPU; and if the VM still fails to start, open a bug.

Reviewer 2: ```suggestion
If the command returns output, note the PID holding the device node of the GPU being prepared, identify the owning service, and stop those services. If no process is holding the GPU; and if the VM still fails to start, open a bug.
```

Example:

NodePrepareResources failure:
```
  Warning  FailedPrepareDynamicResources  22s   kubelet            Failed to prepare dynamic resources: prepare dynamic resources: NodePrepareResources: rpc error: code = DeadlineExceeded desc = context deadline exceeded
```
You can find that the GPU driver switch is being blocked by finding this in the kernel logs:

FInd which process is holding the device node of the GPU being prepared for your workload by parsing the output of the command below and stop the culprit service:
```
$ for f in /proc/[0-9]*/fd/*; do t=$(readlink "$f" 2>/dev/null) || continue; case "$t" in */dev/nvidia[0-9]) echo "PID $(echo "$f" | cut -d/ -f3) holds $t";; esac; done
PID 1210233 holds /run/nvidia/driver/dev/nvidia0
PID 1237013 holds /run/nvidia/driver/dev/nvidia1
```
The subsequent invocation of `NodePrepareResources` for the workload should be successful and it should go to `Running` state.

^^ I've posted an example here at the end. But please integrate this with the whole section above

Author: Done

`DeviceMetadata` requires `PassthroughSupport`, and some gates are mutually exclusive. See the [feature gates reference](../reference/feature-gates.md) and [constraints](../reference/feature-gates.md#constraints) for details.

## Prerequisites

- Meet the general driver [prerequisites](../prerequisites.md).
- **IOMMU enabled** on GPU nodes. VFIO passthrough requires IOMMU; the GPU kubelet plugin fails to start with `PassthroughSupport` enabled if IOMMU is off.
- Use **NVIDIA DRA driver v0.4.0 or later**, which supports [KEP-5304](https://github.com/kubernetes/enhancements/issues/5304) device metadata.
- Install the driver with the **`PassthroughSupport`** and **`DeviceMetadata`** feature gates enabled.

When deploying through the GPU Operator, CDI is enabled by default in GPU Operator v25.10.0 and later. See the [GPU Operator DRA installation guide](https://docs.nvidia.com/datacenter/cloud-native/gpu-operator/latest/dra-intro-install.html).

## Installing the NVIDIA DRA driver

Install the driver with passthrough support and device metadata:

```bash
helm upgrade -i dra-driver-nvidia-gpu oci://registry.k8s.io/dra-driver-nvidia/charts/dra-driver-nvidia-gpu \

>     --version 0.4.0 \

Reviewer 3: ```suggestion
    --version  {{< param "driver_version" >}} \
```

Reviewer 3: we've added in a variable to handling the driver version in code samples so we don't need to manually update this every time. 

Some details on the variable [here](https://dra-driver-nvidia-gpu.sigs.k8s.io/contribute/docs/#updating-docs-for-a-release)

            kind: VfioDeviceConfig
            iommu:
              backendPolicy: LegacyOnly
              enableAPIDevice: true
      requests:
      - name: dra-gpu
        exactly:
          allocationMode: ExactCount
          count: 1
          deviceClassName: vfio.gpu.nvidia.com
```

Save it as `vfio-gpu-rct.yaml`.

### Apply the manifest

```bash
kubectl apply -f vfio-gpu-rct.yaml
```

Example output:

```
resourceclaimtemplate.resource.k8s.io/dra-gpu-claim-template created
```

### VfioDeviceConfig parameters

The opaque `VfioDeviceConfig` block tells the NVIDIA DRA driver which VFIO device nodes to mount into virt-launcher through CDI.

- **`enableAPIDevice: true`** — Mounts the VFIO control device `/dev/vfio/vfio` into the virt-launcher pod. KubeVirt **requires** this device to manage VFIO PCI assignments through libvirt.

- **`backendPolicy: LegacyOnly`** — Selects the legacy IOMMU VFIO backend (`/dev/vfio/<iommu-group>`). The alternative, `PreferIommuFD`, uses the IOMMUFD backend (`/dev/vfio/devices/vfio*`) when available on the host.

Keep `backendPolicy: LegacyOnly` for KubeVirt, which does not support the IOMMUFD backend yet.

## Troubleshooting

> If a VM fails to start, check the kubelet-plugin logs for prepare errors:

Reviewer 2: ```suggestion
If a VM fails to start because its virt-launcher pod is stuck in `ContainerCreating` state, check the kubelet-plugin logs for prepare errors:
```

```bash
kubectl logs -n dra-driver-nvidia-gpu -l dra-driver-nvidia-gpu-component=kubelet-plugin -c gpus
```

A `NodePrepareResources` failure looks something like this:

```
Warning  FailedPrepareDynamicResources  22s   kubelet  Failed to prepare dynamic resources: prepare dynamic resources: NodePrepareResources: rpc error: code = DeadlineExceeded desc = context deadline exceeded
```

> A `DeadlineExceeded` error for `NodePrepareResources` for the workload usually means a service still holds the GPU. On the GPU node, find which process holds a handle on an NVIDIA device:

Reviewer 2: ```suggestion
A `DeadlineExceeded` error for `NodePrepareResources` for the workload usually means a service still holds an open handle to the GPU blocking its driver switch. On the GPU node, the culprit process can be identified using:
```

```bash
for f in /proc/[0-9]*/fd/*; do t=$(readlink "$f" 2>/dev/null) || continue; case "$t" in /dev/nvidia[0-9]*) echo "PID $(echo "$f" | cut -d/ -f3) holds $t";; esac; done
```

Example output:

```
PID 1210233 holds /run/nvidia/driver/dev/nvidia0
PID 1237013 holds /run/nvidia/driver/dev/nvidia1
```

> If the command returns output, note the PID holding the device node of the GPU being prepared, identify the owning service, and stop those services.

Reviewer 2: ```suggestion
If the command returns an output, note the PID holding the device node of the GPU being prepared, identify the owning service, and stop those services.
```
